src/vsr/client_replies.zig

```diff
@@ -0,0 +1,342 @@
+                });
+
+                read.callback(client_replies, null);
+                return;
+            }
+
+            assert(read.message.header.command == .reply);
+            assert(read.message.header.cluster == read.header.cluster);
+
+            if (!read.message.header.valid_checksum_body(read.message.body())) {
+                log.warn("{}: read_reply: corrupt body (client={} reply={})", .{
+                    client_replies.replica,
+                    read.header.client,
+                    read.header.checksum,
+                });
+
+                read.callback(client_replies, null);
+                return;
+            }
+
+            log.debug("{}: read_reply: done (client={} reply={})", .{
+                client_replies.replica,
+                read.header.client,
+                read.header.checksum,
+            });
+
+            read.callback(client_replies, read.message);
+        }
+
+        /// Call `callback` when ClientReplies is able to start another write_reply().
+        pub fn ready(
+            client_replies: *ClientReplies,
+            callback: fn (client_replies: *ClientReplies) void,
+        ) void {
+            assert(client_replies.ready_callback == null);
+            client_replies.ready_callback = callback;
+
+            if (client_replies.writes.available() > 0) {
+                client_replies.storage.on_next_tick(
+                    ready_next_tick_callback,
+                    &client_replies.ready_next_tick,
+                );
+            } else {
+                // ready_callback will be called the next time a write completes.
+            }
+        }
+
+        fn ready_next_tick_callback(next_tick: *Storage.NextTick) void {
+            const client_replies = @fieldParentPtr(ClientReplies, "ready_next_tick", next_tick);
+
+            if (client_replies.ready_callback) |callback| {
+                assert(client_replies.writes.available() > 0);
+
+                client_replies.ready_callback = null;
+                callback(client_replies);
+            } else {
+                // Another write finished before the next_tick triggered.
+            }
+        }
+
+        pub fn write_reply(client_replies: *ClientReplies, slot: Slot, message: *Message) void {
+            assert(message.header.command == .reply);
+            // There is never any need to write a body-less message, since the header is
+            // stored safely in the client sessions superblock trailer.
+            assert(message.header.size != @sizeOf(vsr.Header));
+
+            const write = client_replies.writes.acquire().?;
+            write.* = .{
+                .client_replies = client_replies,
+                .completion = undefined,
+                .message = message.ref(),
+                .slot = slot,
+            };
+
+            // TODO Optimization — if there is already a write for this slot queued (but not started), replace it.
```

Comment:
Interestingly, there's more to this than just avoiding redundant writes. If I am reading the code correctly, then, if _some_ client gets two writes in the write_queue, then all _other_ clients would be blocked, because we won't go past this laggy write in the 


```
            while (client_replies.write_queue.head()) |write| {
                if (client_replies.writing.isSet(write.slot.index)) break;
```

loop. 

This actually seems pretty bad, as it creates two equilibrium states. Normally, we can serve `client_replies_iops_write_max` clients in concurrently. 

Now, if one of those clients bursts and causes two writes in the queue, all other clients stall. So it becomes likely that some other cilent queues two writes. And that can lock us into usining only one write iop, despite the fact that we have a bunch of enqueued writes, and a bunch of iops available.

Reply:
:+1:  I'll implement this TODO. But if the blocking does end up as a perf issue, we can switch out the ring buffer queue for a linked list. (I just used a ring buffer for simplicity.)